### ISSUE TITLE
[tracking issue] warnings on long running constant evaluations

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#49980]
https://github.com/rust-lang/rust/pull/49947 introduces a warning (removing the previous hard error).

This warning is currently unsilenceable and should probably become a lint. For simplicity (https://github.com/rust-lang/rust/pull/49947 needs to be backported to beta) this has not been done.

"When" the warning appears is deterministic, but depends on MIR, so it might change between versions of rustc.

**Comment by rcoh** [comment]
Seems like this will also require a small amount of refactoring? I don't think the evaluator where the warning is currently emitted knows the `NodeId` that is needed when emitting the lint.

**Comment by oli-obk** [comment]
We should be reporting the lint on the constant that is being evaluated, not the place that we are currently at, because that place might change between rustc versions

**Comment by oli-obk** [comment]
Note that just turning this into a lint is probably not enough. We need to also check whether the lint is `deny` in the constant's scope. If it is, we need to abort the interpretation. Otherwise we'd just keep getting these lints as errors without ever stopping compilation.

**Comment by steveklabnik** [comment]
Triage: I'm not aware of any changes here. I'm also not sure what to do to reproduce this issue, to be honest.

**Comment by oli-obk** [comment]
`static FOO: () = loop {};` gives (on beta) 

```
error[E0080]: could not evaluate static initializer
 --> src/lib.rs:1:18
  |
1 | static FOO: () = loop {};
  |                  ^^^^^^^ exceeded interpreter step limit (see `#[const_eval_limit]`)
```

So I think we're good here
